vending-app/src/main/java/org/microg/vending/billing/core/HttpClient.kt

```diff
@@ -10,14 +10,47 @@ import com.squareup.wire.Message
 import com.squareup.wire.ProtoAdapter
 import org.json.JSONObject
 import org.microg.gms.utils.singleInstanceOf
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
 import kotlin.coroutines.resume
 import kotlin.coroutines.resumeWithException
 import kotlin.coroutines.suspendCoroutine
 
 private const val POST_TIMEOUT = 8000
 
 class HttpClient(context: Context) {
-    private val requestQueue = singleInstanceOf { Volley.newRequestQueue(context.applicationContext) }
+
+    val requestQueue = singleInstanceOf { Volley.newRequestQueue(context.applicationContext) }
+
+    suspend fun download(url: String, downloadFile: File, tag: String): String = suspendCoroutine { continuation ->
+        val uriBuilder = Uri.parse(url).buildUpon()
+        requestQueue.add(object : Request<String>(Method.GET, uriBuilder.build().toString(), null) {
+            override fun parseNetworkResponse(response: NetworkResponse): Response<String> {
+                if (response.statusCode != 200) throw VolleyError(response)
+                return try {
+                    val parentDir = downloadFile.getParentFile()
+                    if (parentDir != null && !parentDir.exists() && !parentDir.mkdirs()) {
+                        throw IOException("Failed to create directories: ${parentDir.absolutePath}")
+                    }
+                    val fos = FileOutputStream(downloadFile)
+                    fos.write(response.data)
```

Comment:
Volley load the entire response into memory. This is a problem for large downloads, as it will consume a lot of RAM (temporarily). The data from the network stream should be written to disk immediately instead.